### PR TITLE
goto_symex: implement SET_RETURN_VALUE

### PR DESCRIPTION
**Pull request by kroening**
- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-programs/mm_io.cpp [hunk 1]
```diff
@@ -105,26 +106,37 @@ void mm_io(
...
+      to_code_type(mm_io_r.type()).return_type(),
+      id2string(id_r) + "$value",
+      id2string(id_r) + "$value",
+      maybe_symbol->location,
+      maybe_symbol->mode,
+      symbol_table);
+
+    mm_io_r_value = value_symbol.symbol_expr();
+  }
+
```
Comment (tautschnig) [inline src/goto-programs/mm_io.cpp:133]: Why are the changes to `mm_io` necessary?

File: src/goto-programs/process_goto_program.cpp [hunk 2]
```diff
@@ -58,9 +58,16 @@ bool process_goto_program(
   }
 
   // remove returns, gcc vectors, complex
-  remove_returns(goto_model);
+  if(
+    options.get_bool_option("remove-returns") ||
+    options.get_bool_option("string-abstraction"))
```
Comment (tautschnig) [inline src/goto-programs/process_goto_program.cpp:63]: String abstraction was built well before return-value removal, so I wouldn't really expect a dependency here. In 425731b66727c40607457cef1a82acb3f6f316b6 I noticed a need for some fixes, but I don't think that there should be a general requirement for return-value removal to make string abstraction work.

File: src/goto-programs/validate_goto_model.cpp [hunk 3]
```diff
@@ -48,7 +48,7 @@ class validate_goto_modelt
   /// - no return statements in any of the functions
   /// - lhs of every \ref code_function_callt instruction is nil
   /// - all return types are void (of both calls and functions themselves)
-  void check_returns_removed();
+  // void check_returns_removed();
```
Comment (tautschnig) [inline src/goto-programs/validate_goto_model.cpp]: I think we should either completely remove this check, or else introduce a bit map for goto programs that indicates which passes have run (and then make this check conditional).

File: src/goto-programs/mm_io.cpp (hunk 1 above)
Comment (kroening) [reply src/goto-programs/mm_io.cpp:133]: The current implementation assumes that the return value is in the function#return_value variable. It needs to be changed to use the lhs in the function call.

File: src/goto-programs/process_goto_program.cpp (hunk 2 above)
Comment (kroening) [reply src/goto-programs/process_goto_program.cpp:63]: It's unclear what the effort is. Might not be worth it, given that this feature has no known users.

File: src/goto-programs/validate_goto_model.cpp (hunk 3 above)
Comment (kroening) [reply src/goto-programs/validate_goto_model.cpp]: Ok, will remove.

Comment (kroening) [reply src/goto-programs/validate_goto_model.cpp]: done

**Review by tautschnig — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**